spdx/v3/v3_0/embedded_first/model.go

```diff
@@ -0,0 +1,161 @@
+package v3_0
+
+import "time"
+
+type ElementID string
+
+// ----------------------- SpdxDocument -----------------------
+
+type SpdxDocument struct {
+	ElementCollection
+
+	NamespaceMap map[string]string
+}
+
+type ISpdxDocument interface {
+	IElementCollection
+}
+
+// ----------------------- ElementCollection -----------------------
+
+type ElementCollection struct {
+	Element
+
+	RootElements []IElement
+	Elements     []IElement
```

Comment:
So the problem with this implementation is that RootElement is not class Element but any class that inherits from Element. In Go, I guess, that translates to "any struct that implements the Element interface".

Reply:
Yes, in this variant, the `I*` types are all `interfaces` -- so these are "any types implementing the `IElement` interface". These could be named differently... naming is hard! there doesn't seem to be a way to avoid interfaces unless we made any reference field an `any` type, but then there's no compiler enforcement of elements going in there. (Also I probably modeled this wrong, I was going by the [software profile svg](https://raw.githubusercontent.com/spdx/spdx-3-model/main/model%20Software.svg) which appears to have `*` for both of these fields).

Reply:
FWIW, I _think_ personally I'm leaning towards the `embedded_with_interfaces_and_constructors` variant, with setters that return `error`, [as the example currently shows](https://github.com/spdx/tools-golang/pull/240/files/c8d843ccbcda6cf5128a628e42da75a53601d228#diff-1f2cba19eca292de24e0f81517e713ff958d56252243c2c4192359fdf2aaa0d3R13). This allows generated code to enforce valid values for all fields and for a user to most quickly understand when things failed rather than needing to run a full validation once all values are set. It also eliminates the need to export the data structs (all the embedded variants need exported structs)

Reply:
Oh my bad: the property is of type IElement

Reply:
This seems to work. It would be easier to generate:
```
package main

import "fmt"

type ElementP struct {
	Name string
}

type Element interface {
	AsElement() ElementP
}

type Artifact struct {
	ElementP
	StandardName []string
}

func (a Artifact) AsElement() ElementP {
	return a.ElementP
}

type ElementCollection struct {
	ElementP
	RootElement Element
}

func main() {
	a := Artifact{
		ElementP: ElementP{
			Name: "AnArtifact",
		},
		StandardName: []string{"TheArtifact"},
	}

	ec := ElementCollection{
		ElementP: ElementP{
			Name: "AnElementCollection",
		},
		RootElement: a,
	}

	fmt.Println(ec)
}
```

Reply:
One more issue: The jinja renderer I am using from the shacl2code project can only generate one file. So if we were to generate a model and a constructor file, we would need to roll our own jinja renderer...which I would rather do once I can handle the code better. You are welcome to try building one that will generate multiple files.

Reply:
One note: there is no requirement to have separate files -- I just did this to separate things logically, but all the code for each of these approaches could be generated in a single file. In other words: if we were to export all 4 types: interface, data struct, constructor, constructor props; all of these could be one after the other in the same file.

From the "generic shacl2code" perspective, I think in the example there's a bit of an inconsistency with the example you posted: only a single struct type gets renamed (`ElementP`) for an `Element` interface, but there are no other interfaces created. I think the easiest thing to do in `shacl2code` is probably: generate an exported interface and a struct for each type. It also probably makes the most sense to name the _interface_ after the shacl type rather than the struct, as one of my examples has. Abstract types are a question: golang would let you indiscriminately construct any abstract type if these structs are exported, and they would need to be generated and exported to be used like superclasses through embedding.

The `interfaces_only` approach has the potential to model inheritance defined in shacl in the most correct way, I believe -- because abstract classes can _omit_ constructor functions and since the types themselves are not exported, there's no way for a user to create these types incorrectly. It has some complexity that at the time you're generating code for a type, it needs to have access to supertypes to be able to inline all the properties -- a quick look makes me think this wouldn't be too hard to accomplish, since it looks like we can [get a reference to a parent class](https://github.com/JPEWdev/shacl2code/pull/9/files#diff-3d82489614fc2a7b7f166bce8660b271d2224754c390411135fd294ea15c3d4bR92) and [jinja supports recursion](https://stackoverflow.com/questions/32771924/macro-recursion-in-jinja2).

An aside: _naming is hard_! Is the `P` from `ElementP` short for `Props` or `Properties`? I am sorta partial to `Data` for these structs despite using `Impl` in one of the variants, but also like `Props`, at least for the constructor props types, or other things; happy for any other suggestions but would definitely like to keep it short but maybe more descriptive than a single letter. Interfaces only allows the lowercase name to be used, since it's not exported and we don't have to come up with names, so I guess there's another plus for that one...

Reply:
I think I agree with you on going with the interface only approach i.e. Abstract class -> Go Interface with getter and setter for each property. Concrete class -> Go struct which implements the getter and setter methods for the abstract class.

I personally would prefer having corresponding internal structs to hold common properties in the concrete classes, but it's not a deal breaker. We have to embed the interfaces anyway (Artifact is actually an abstract class that inherits from Element).